plugin/dns64 : add support for DNS requests over IPv4 network (#4809)
miekg/coredns: plugin/dns64/README.md

dns64 [PREFIX] {
    [translate_all]
    prefix PREFIX
    [allow_ipv4]
}
~~~

* `prefix` specifies any local IPv6 prefix to use, instead of the well known prefix (64:ff9b::/96)
* `translate_all` translates all queries, including responses that have AAAA results.

> * `allow_ipv4` allow DNS queries, ignoring whether they came from over IPv4 or IPv6 network

Reviewer: Similar here, let's simplify the wording.

```suggestion
* `allow_ipv4` Allow translating queries if they come in over IPv4, default is IPv6 only translation.
```

Author: yeah, that sounds much better 👍  fixed as you suggested

}
~~~

> Apply translation even to the requests which arrived over IPv4 network.

Reviewer: Can we add a note that this may break `A` record lookup for users coming from dual-stack networks?

Author: added a warning, hope the wording is understandable 👍

}
~~~

Apply translation even to the requests which arrived over IPv4 network. Beware that enabling this option in CoreDNS handling combination 

> of dual-stack and IPv6 clients might cause problems by also doing DNS64 translations for requests of dual-stack clients.

Reviewer 2: ... but the plugin will do translations for dual-stack clients even if the option is not enabled. So, the behavior described is not specific to this option.  As long as a dual-stack client makes a query for a `AAAA` over IPv6 transport, this plugin will attempt a translation (when there are no `AAAA` records).

The difference In behavior would be for `AAAA` queries made over IPv4 transport.  By default, the plugin gives a different answer to the client depending on the transport version. On the surface that seems potentially problematic - different answers for same name/type depending on which IP you send the request out on.

The `allow_ipv4` option makes the answer consistent for both transports.

Reviewer: How about this wording:

```
Warning, the `allow_ipv4` feature will apply translations to requests coming from dual-stack clients. This means that a request for a client that sends an `AAAA` that would normal result in an `NXDOMAIN` would get a translated result. This may cause unwanted IPv6 dns64 traffic when a dualstack client would normally use the result of an `A` record request.
```

@chrisohaver Does this better explain it/warn the user?

Reviewer 2: fwiw, I don’t understand why it’s a warning. It seems like that is how it should behave by default. But if you are ok with it then let’s go with that.

Reviewer: It's a warning because users with functioning IPv4 will get re-routed to IPv6 unnecessarily. That's why we implemented it the current way by default. This plugin is meant for users who are IPv6-only, not dualstack. If the user can reach IPv4, they should get untranslated results.

Author: I have adjusted wording as you suggested @SuperQ, please take a look if it makes sense

}
~~~

Apply translation even to the requests which arrived over IPv4 network. Beware that enabling this options may break dual-stack clients,

> by doing DNS64 translations for empty AAAA records (no IPv6 for the domain) and clients might prefer using synthetic records instead of using A records.

Reviewer 2: Wouldn't this happen even without the option enabled?

Author: ummm, right. Bad wording :D

Author: hope it makes sense now, I tried to describe the example the @SuperQ gave in the comments
